### PR TITLE
Refactor conv warnings

### LINKED ISSUE #2615: [FEATURE] Issue an warning if PollHandler (etc.?) is added into a conv handler
We warn for quite some handlers if the surrounding variables dont work out for them (CallbackQuery, ChosenInlineResultHandler). We can at least add Poll and PollAnswerHandler to it, maybe some more I missed. We can maybe even write some clearer error messages. Suggestions are welcome :)

### PR DESCRIPTION
**Pull request by Poolitzer**
### Checklist for PRs

- [ ] Added `.. versionadded:: version`, `.. versionchanged:: version` or `.. deprecated:: version` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)
- [x] Created new or adapted existing unit tests
- [ ] Added myself alphabetically to `AUTHORS.rst` (optional)

This PR closes #2615.

I did make a decision while implementing this: I don't think it makes sense to break the for loops after one warning. The warnings menu will not show more then one unique warning anyway, and we swallow non unique ones this way or unique ones if the user changes their `filterwarning` method.

### CODE REVIEW

File: telegram/ext/_conversationhandler.py [hunk 1]
```diff
@@ -283,35 +293,65 @@ def __init__(
...
+        # this link will be added to all warnings tied to per_* setting
+        per_faq_link = (
+            " Read this FAQ entry to learn more about the per_* settings https://git.io/JtcyU."
+        )
+
+        for handler in all_handlers:
+            if isinstance(handler, (StringCommandHandler, StringRegexHandler)):
+                warn(
+                    "The ConversationHandler does not work with non Telegram.Update type updates, "
+                    "and you shouldn't use this handler for Telegram Update types.",
```
Comment (Bibo-Joshi) [inline telegram/ext/_conversationhandler.py:308]:
```suggestion
                    "The `ConversationHandler` only handles updates of type `telegram.Update`. "
                    f"{handler.__class__.__name__} handles updates of type `str`.",
```

File: telegram/ext/_conversationhandler.py [hunk 2]
```diff
@@ -283,35 +293,65 @@ def __init__(
...
+                    stacklevel=2,
+                )
+            if isinstance(handler, PollHandler):
+                warn(
+                    "PollHandler will never trigger in a conversation since it has no information "
+                    "about the chat or the user who voted in it. Do you mean the "
+                    "'PollAnswerHandler'?",
```
Comment (Bibo-Joshi) [inline telegram/ext/_conversationhandler.py:315]:
```suggestion
                    "`PollAnswerHandler`?",
```

File: telegram/ext/_conversationhandler.py [hunk 3]
```diff
@@ -283,35 +293,65 @@ def __init__(
...
+                    handler,
+                    (
+                        ShippingQueryHandler,
+                        InlineQueryHandler,
+                        ChosenInlineResultHandler,
+                        PreCheckoutQueryHandler,
+                        PollAnswerHandler,
+                    ),
+                )
+                and self.per_chat
```
Comment (Bibo-Joshi) [inline telegram/ext/_conversationhandler.py:332]: can you swap this to `if self.per_chat and isinstance(…)`?

File: telegram/ext/_conversationhandler.py [hunk 4]
```diff
@@ -283,35 +293,65 @@ def __init__(
...
+            ):
+                warn(
+                    "This Handler only has information about the user, so it wont ever be"
```
Comment (Bibo-Joshi) [inline telegram/ext/_conversationhandler.py:335]:
```suggestion
                    f"Updates handled by {handler.__class__.__name__} only have information about the user, so this handler wont ever be"
```

File: telegram/ext/_conversationhandler.py [hunk 5]
```diff
@@ -283,35 +293,65 @@ def __init__(
...
+                    " triggered if 'per_chat=True'." + per_faq_link,
```
Comment (Bibo-Joshi) [inline telegram/ext/_conversationhandler.py:336]:
```suggestion
                    " triggered if `per_chat=True`." + per_faq_link,
```

File: telegram/ext/_conversationhandler.py [hunk 6]
```diff
@@ -283,35 +293,65 @@ def __init__(
...
+        # this loop is going to warn the user about handlers which can work unexpected
+        # in conversations
+
+        # this link will be added to all warnings tied to per_* setting
+        per_faq_link = (
+            " Read this FAQ entry to learn more about the per_* settings https://git.io/JtcyU."
+        )
+
+        for handler in all_handlers:
+            if isinstance(handler, (StringCommandHandler, StringRegexHandler)):
```
Comment (Bibo-Joshi) [inline telegram/ext/_conversationhandler.py:305]: what about `TypeHandler` with non-TG-update type?

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the PR :) I've left a couple of nitpicks. In addition to that, could you add an item to the PR template about adding new handlers to these checks?

File: telegram/ext/_conversationhandler.py [hunk 7]
```diff
@@ -283,35 +293,65 @@ def __init__(
...
+            if isinstance(handler, PollHandler):
+                warn(
+                    "PollHandler will never trigger in a conversation since it has no information "
+                    "about the chat or the user who voted in it. Do you mean the "
+                    "'PollAnswerHandler'?",
+                    stacklevel=2,
+                )
+            # we can assume per_user is set. If its not, per_message has to be set, otherwise the
+            # user would have faced the ValueError that all per_* settings cant be false. If
+            # per_message is set, the user will get a warning later.
```
[the remark this answers is not in the document]
Comment (Poolitzer) [reply telegram/ext/_conversationhandler.py:320]: I typed out a long ass respond why I am correct but then I realized that explicitly having the check there is better future proof. so Im doing that.

File: telegram/ext/_conversationhandler.py [hunk 8]
```diff
@@ -283,35 +294,73 @@ def __init__(
...
+                            ChosenInlineResultHandler,
+                            PreCheckoutQueryHandler,
+                            PollAnswerHandler,
+                        ),
+                    )
+                )
+            ):
+                warn(
+                    f"Updates handled by {handler.__class__.__name__} only have information about "
+                    f"the user, so this handler wont ever be triggered if `per_chat=True`."
```
Comment (harshil21) [inline telegram/ext/_conversationhandler.py:344]:
```suggestion
                    f"the user, so this handler won't ever be triggered if `per_chat=True`."
```

File: telegram/ext/_conversationhandler.py [hunk 9]
```diff
@@ -283,35 +294,73 @@ def __init__(
...
+                    stacklevel=2,
+                )
+
+            if self.per_message:
                 if not isinstance(handler, CallbackQueryHandler):
                     warn(
                         "If 'per_message=True', all entry points, state handlers, and fallbacks"
                         " must be 'CallbackQueryHandler', since no other handlers "
-                        "have a message context.",
+                        "have a message context." + per_faq_link,
```
Comment (harshil21) [inline telegram/ext/_conversationhandler.py:354]:
```suggestion
                        "have a message context.{per_faq_link}",
```
how come pylint didn't nag about f-string use? Same below and above

File: telegram/ext/_conversationhandler.py [hunk 10]
```diff
@@ -283,35 +294,73 @@ def __init__(
...
             all_handlers.extend(state_handlers)
 
-        if self.per_message:
-            for handler in all_handlers:
+        # this loop is going to warn the user about handlers which can work unexpected
+        # in conversations
+
+        # this link will be added to all warnings tied to per_* setting
+        per_faq_link = (
+            " Read this FAQ entry to learn more about the per_* settings https://git.io/JtcyU."
```
Comment (harshil21) [inline telegram/ext/_conversationhandler.py:302]:
```suggestion
            " Read this FAQ entry to learn more about the per_* settings: https://git.io/JtcyU."
```

File: telegram/ext/_conversationhandler.py [hunk 11]
```diff
@@ -283,35 +294,73 @@ def __init__(
...
+        )
+
+        for handler in all_handlers:
+            if isinstance(handler, (StringCommandHandler, StringRegexHandler)):
+                warn(
+                    "The `ConversationHandler` only handles updates of type `telegram.Update`. "
+                    f"{handler.__class__.__name__} handles updates of type `str`.",
+                    stacklevel=2,
+                )
+            if isinstance(handler, TypeHandler) and handler.type is not Update:
```
Comment (harshil21) [inline telegram/ext/_conversationhandler.py:312]: these ifs can be converted to elifs

File: tests/test_conversationhandler.py [hunk 12]
```diff
@@ -1325,112 +1334,137 @@ def slowbrew(_update, context):
         assert handler.conversations.get((self.group.id, user1.id)) is None
         assert self.is_timeout
 
-    def test_conversation_timeout_warning_only_shown_once(self, recwarn):
+    def test_handlers_generate_warning(self, recwarn):
+        # this function tests all handler + per_* setting combinations.
```
Comment (harshil21) [inline tests/test_conversationhandler.py:1338]: convert to a docstring instead?

**Review by harshil21 — CHANGES_REQUESTED** [review]
couple of tiny changes-

File: telegram/ext/_conversationhandler.py (hunk 9 above)
Comment (Bibo-Joshi) [reply telegram/ext/_conversationhandler.py:354]: I think it will only complain about `.format`. though f-strings seem to be [a bit faster](https://stackoverflow.com/a/38722192/10606962). I'm okay with both here …

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the updates :) just a bit more nitpicking …

File: telegram/ext/_conversationhandler.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (harshil21) [reply telegram/ext/_conversationhandler.py:312]: Shouldn't it be `and not issubclass(..)`? 

Also that part should be added to the tests, i.e TypeHandler with a subclass of Update should not raise a warning

File: telegram/ext/_conversationhandler.py [hunk 13]
```diff
@@ -308,32 +309,40 @@ def __init__(
...
-                    handler,
-                    (
-                        ShippingQueryHandler,
-                        InlineQueryHandler,
-                        ChosenInlineResultHandler,
-                        PreCheckoutQueryHandler,
-                        PollAnswerHandler,
-                    ),
+                self.per_chat
+                and self.per_user
```
[the remark this answers is not in the document]
Comment (Poolitzer) [reply telegram/ext/_conversationhandler.py:328]: fair enough. I misinterpreted your [previous comment](https://github.com/python-telegram-bot/python-telegram-bot/pull/2755#discussion_r739639248) then. Not sure why I thought this is a good idea though

File: .github/pull_request_template.md [hunk 14]
```diff
@@ -28,6 +28,7 @@ Hey! You're PRing? Cool! Please have a look at the below checklist. It's here to
     - [ ] Link new and existing constants in docstrings instead of hard coded number and strings
     - [ ] Add new message types to `Message.effective_attachment`
     - [ ] Added new handlers for new update types
+      - [ ] Add the handlers to the warning loop in the `ConversationHandler`
```
[the remark this answers is not in the document]
Comment (Poolitzer) [reply .github/pull_request_template.md:31]: my intention was to indent it, because it is only needed if new handlers are added. Apparently I need to do two spaces, so thats wrong here anyway, but I would still indent it.

File: telegram/ext/_conversationhandler.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Poolitzer) [reply telegram/ext/_conversationhandler.py:312]: it is: https://github.com/python-telegram-bot/python-telegram-bot/blob/5bf1d14e39e238ba99f2a04e866428a9b0db7480/tests/test_conversationhandler.py#L1357-L1360

File: .github/pull_request_template.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply .github/pull_request_template.md:31]: ah okay. good point 👍

**Review by Bibo-Joshi — APPROVED** [review]
nice addition 👍  if @harshil21 gives a LGTM as well, we can merge :)

File: telegram/ext/_conversationhandler.py [hunk 15]
```diff
@@ -283,35 +294,67 @@ def __init__(
...
+                )
+            elif not self.per_message and isinstance(handler, CallbackQueryHandler):
+                warn(
+                    "If 'per_message=False', 'CallbackQueryHandler' will not be "
+                    f"tracked for every message.{per_faq_link}",
+                    stacklevel=2,
+                )
 
         if self.conversation_timeout:
             for handler in all_handlers:
```
Comment (harshil21) [inline telegram/ext/_conversationhandler.py:360]: Is it okay to move this conversation timeout warning inside the first for loop? Unnecessary to iterate twice over the handlers just to warn again imo

File: telegram/ext/_conversationhandler.py [hunk 16]
```diff
@@ -322,7 +365,6 @@ def __init__(
                         "differently from what you expect.",
                         stacklevel=2,
                     )
-                    break
 
         if self.run_async:
             for handler in all_handlers:
```
Comment (harshil21) [inline telegram/ext/_conversationhandler.py:370]: same thing here, do this in one pass

File: tests/test_conversationhandler.py [hunk 17]
```diff
@@ -1325,112 +1334,139 @@ def slowbrew(_update, context):
...
-            conversation_timeout=100,
-        )
-        assert len(recwarn) == 1
-        assert str(recwarn[0].message) == (
-            "Using `conversation_timeout` with nested conversations is currently not "
-            "supported. You can still try to use it, but it will likely behave "
-            "differently from what you expect."
+            fallbacks=[TypeHandler(Update, self.code)],
         )
-        assert recwarn[0].filename == __file__, "incorrect stacklevel!"
```
Comment (harshil21) [inline tests/test_conversationhandler.py:1360]: why was this removed but not added back? If it was a mistake, then you can use a for loop to iterate over warnings and then check it.

**Review by harshil21 — COMMENTED** [review]
some more optimizations-

File: tests/test_conversationhandler.py (hunk 17 above)
Comment (Poolitzer) [reply tests/test_conversationhandler.py:1360]: Im not sure what this assert checks, I can readd it.

Also not sure what you meant with for loop, do you want to go over all recwarnings and assert the filename?

Comment (harshil21) [reply tests/test_conversationhandler.py:1360]: > Im not sure what this assert checks, I can readd it.

It checks if you passed in the correct `stacklevel` parameter to `warn()`. It does this by checking if the filename of the warning emitted matches the filename of the current module.

> Also not sure what you meant with for loop, do you want to go over all recwarnings and assert the filename?

yep exactly that

File: telegram/ext/_conversationhandler.py [hunk 18]
```diff
@@ -366,8 +365,7 @@ def __init__(
                         stacklevel=2,
                     )
 
-        if self.run_async:
-            for handler in all_handlers:
+            elif self.run_async:
```
Comment (harshil21) [inline telegram/ext/_conversationhandler.py:368]: this should be an `if` right? Just to be safe..

Comment (Poolitzer) [reply telegram/ext/_conversationhandler.py:368]: both that and timeout, yeah. Those have nothing to do with the previous ifs.

**Review by harshil21 — APPROVED** [review]
Good job!

**Comment by Bibo-Joshi** [comment]
Tests failing due to API changes. Merging :)
